lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -1091,292 +544,461 @@ function Enable-Service {
+        return @{}
+    }
 
-function Stop-Neutron {
-    $services = Get-CharmServices
-    $net_type = Get-NetType
-    if ($net_type -eq "hyperv"){
-        Stop-Service $services["neutron"]["service"]
-    } elseif ($net_type -eq "ovs") {
-        Stop-Service $services['neutron-ovs']['service']
+    $required = @{
+        "glance-api-server"=$null;
+    }
+    $ctx = Get-JujuRelationContext -Relation 'image-service' -RequiredContext $required -OptionalContext $optionalCtx
+    $new = @{}
+    foreach ($i in $ctx.Keys) {
+        $new[$i.Replace("-", "_")] = $ctx[$i]
     }
+    return $new
 }
 
-function Import-CloudbaseCert {
-    $filesDir = Get-FilesDir
-    $crt = Join-Path $filesDir "Cloudbase_signing.cer"
-    if (!(Test-Path $crt)){
-        return $false
+function Get-CharmConfigContext {
+    $config = Get-JujuCharmConfig
+    $asHash = @{}
+    foreach ($i in $config.GetEnumerator()) {
+        $name = $i.Key
+        if ($name -eq "instances-dir") {
+            continue
+        }
+        if ($i.Value.Gettype() -is [System.String]) {
+            $v = ($i.Value).Replace('/', '\')
+        } else {
+            $v = $i.Value
+        }
+        $asHash[$name.Replace("-", "_")] = $v 
     }
-    Import-Certificate $crt -StoreLocation LocalMachine -StoreName TrustedPublisher
+    [string]$ip = Get-JujuUnitPrivateIP
+    $asHash['my_ip'] = $ip 
+    return $asHash
 }
 
-function Start-WSFCRelationJoinedHook {
-    $ctx = Get-ActiveDirectoryContext
-    if(!$ctx.Count -or !(Confirm-IsInDomain $ctx["domainName"])) {
-        Set-ClusterableStatus -Ready 0 -Relation "failover-cluster"
-        return
+function Get-SystemContext {
+    $asHash = @{
+        "installDir" = $INSTALL_DIR;
+        "force_config_drive" = "False";
+        "config_drive_inject_password" = "False";
+        "config_drive_cdrom" = "False";
     }
-
-    if (Get-IsNanoServer) {
-        $features = @('FailoverCluster-NanoServer')
-    } else {
-        $features = @('File-Services','FailoverCluster-FullServer','FailoverCluster-Powershell')
+    if ((Get-IsNanoServer)) {
+        $asHash["force_config_drive"] = "True";
+        $asHash["config_drive_inject_password"] = "True";
+        $asHash["config_drive_cdrom"] = "True";
     }
-    Install-WindowsFeatures -Features $features
-    Set-ClusterableStatus -Ready 1 -Relation "failover-cluster"
+    return $asHash
 }
 
-function Start-ConfigChangedHook {
-    Start-ConfigureVMSwitch
-    Confirm-ServicePrerequisites
-    Generate-ExeWrappers
-    
-    $net_type = Get-NetType
-
-    if ($net_type -eq "ovs"){
-        $neutron_restart = Generate-Config -ServiceName "neutron-ovs"
-    }else{
-        $neutron_restart = Generate-Config -ServiceName "neutron"
-    }
-
-    $nova_restart = Generate-Config -ServiceName "nova"
-    $JujuCharmServices = Get-CharmServices
-
-    if ($nova_restart){
-        Write-JujuInfo "Restarting service Nova"
-        Restart-Nova
+function Set-IncompleteStatusContext {
```

Comment:
I think this can be simplified. If you have any incomplete contexts just set the proper status. If you don't just say all is well. Much less error prone. Use your best judgment, I just feel this could be cleaner.